### PR TITLE
#164 Update network inputs

### PR DESCRIPTION
**Pull request by lecourtoisflo**

### CODE REVIEW

File: sources/Inputs/include/Node.h [hunk 1]
```diff
@@ -48,6 +48,107 @@ struct VoltageLevel {
...
+   *
+   * @param lineId the line id
+   * @param node1 the origin of the line
+   * @param node2 the extremity of the line
+   */
+  Line(const LineId& lineId, const std::shared_ptr<Node>& node1, const std::shared_ptr<Node>& node2);
+};
+
+/**
+ * @brief topological Transfo
```
Comment (rosiereflo) [inline sources/Inputs/include/Node.h:84]: Transformer (same everywhere)

File: sources/Inputs/include/Node.h [hunk 2]
```diff
@@ -65,14 +166,18 @@ class Node {
...
    * @returns the built node
    */
-  static std::shared_ptr<Node> build(const NodeId& id, const std::shared_ptr<VoltageLevel>& vl, double nominalVoltage);
+  static std::shared_ptr<Node> build(const NodeId& id, const std::shared_ptr<VoltageLevel>& vl, double nominalVoltage, const std::vector<Shunt>& shunts);
 
   const NodeId id;                                      ///< node id
   const std::weak_ptr<VoltageLevel> voltageLevel;       ///< voltage level containing the node
   const double nominalVoltage;                          ///< Nominal voltage of the node
+  const std::vector<Shunt> shunts;                      ///< Shunts connectables to the node
+  std::vector<std::weak_ptr<Line>> lines;               ///< Lines to which the node belong
```
Comment (rosiereflo) [inline sources/Inputs/include/Node.h:179]: "Lines connected to this node" and similarly for transfos

File: sources/Inputs/include/Node.h [hunk 3]
```diff
@@ -65,14 +166,18 @@ class Node {
...
    *
    * @returns the built node
    */
-  static std::shared_ptr<Node> build(const NodeId& id, const std::shared_ptr<VoltageLevel>& vl, double nominalVoltage);
+  static std::shared_ptr<Node> build(const NodeId& id, const std::shared_ptr<VoltageLevel>& vl, double nominalVoltage, const std::vector<Shunt>& shunts);
 
   const NodeId id;                                      ///< node id
   const std::weak_ptr<VoltageLevel> voltageLevel;       ///< voltage level containing the node
   const double nominalVoltage;                          ///< Nominal voltage of the node
+  const std::vector<Shunt> shunts;                      ///< Shunts connectables to the node
```
Comment (rosiereflo) [inline sources/Inputs/include/Node.h:178]: typo connectable (same everywhere)

File: sources/Inputs/src/NetworkManager.cpp [hunk 4]
```diff
@@ -55,68 +56,78 @@ NetworkManager::buildTree() {
   auto opt_id = network->getSlackNodeBusId();
 
   const auto& voltageLevels = network->getVoltageLevels();
-  for (auto it_v = voltageLevels.begin(); it_v != voltageLevels.end(); ++it_v) {
-    auto vl = std::make_shared<VoltageLevel>((*it_v)->getID());
+  for (const auto& networkVL : voltageLevels) {
+    const auto& shunts = networkVL->getShuntCompensators();
+    std::map<Node::NodeId, std::vector<Shunt>> shuntsMap;
```
Comment (rosiereflo) [inline sources/Inputs/src/NetworkManager.cpp:61]: Why a map and not a hashmap?

File: sources/Inputs/src/Node.cpp [hunk 5]
```diff
@@ -67,5 +68,61 @@ operator>=(const Node& lhs, const Node& rhs) {
 
 VoltageLevel::VoltageLevel(const VoltageLevelId& vlid) : id(vlid) {}
 
+/////////////////////////////////////////////////
+
+std::shared_ptr<Line>
+Line::build(const LineId& lineId, const std::shared_ptr<Node>& node1, const std::shared_ptr<Node>& node2) {
+  auto ret = std::shared_ptr<Line>(new Line(lineId, node1, node2));
+  node1->neighbours.push_back(node2);
```
Comment (rosiereflo) [inline sources/Inputs/src/Node.cpp:76]: What if node1 or node2 is nullptr?

File: sources/Inputs/src/Node.cpp [hunk 6]
```diff
@@ -67,5 +68,61 @@ operator>=(const Node& lhs, const Node& rhs) {
 
 VoltageLevel::VoltageLevel(const VoltageLevelId& vlid) : id(vlid) {}
 
+/////////////////////////////////////////////////
+
+std::shared_ptr<Line>
+Line::build(const LineId& lineId, const std::shared_ptr<Node>& node1, const std::shared_ptr<Node>& node2) {
```
Comment (rosiereflo) [inline sources/Inputs/src/Node.cpp:74]: It is very misleading to have node1 and node2 const whereas we edit the referenced value...

File: sources/Outputs/src/Dyd.cpp [hunk 7]
```diff
@@ -161,7 +161,7 @@ Dyd::writeHvdcLine(const algo::HvdcLineDefinition& hvdcLine, const std::string&
   }
 
   return model;
-}  // namespace outputs
```
Comment (rosiereflo) [inline sources/Outputs/src/Dyd.cpp:164]: Why removing this?

File: tests/inputs/TestNode.cpp [hunk 8]
```diff
@@ -28,3 +28,21 @@ TEST(TestNode, base) {
   ASSERT_TRUE(*node2 >= *node);
   ASSERT_TRUE(*node0 >= *node);
 }
+
+TEST(TestNode, line) {
+  auto vl = std::make_shared<dfl::inputs::VoltageLevel>("VL");
+  std::vector<dfl::inputs::Shunt> shunts1 = {dfl::inputs::Shunt("1.1")};
```
Comment (rosiereflo) [inline tests/inputs/TestNode.cpp:34]: Unit tests are missing for at least transfo and network manager with shunts. Please make sure the overall coverage is as high as possible

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Inputs/include/Node.h (hunk 1 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:84]: OK

File: sources/Inputs/include/Node.h (hunk 2 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:179]: OK

File: sources/Inputs/include/Node.h (hunk 3 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:178]: OK

File: sources/Inputs/src/NetworkManager.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply sources/Inputs/src/NetworkManager.cpp:61]: OK

File: sources/Inputs/src/Node.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Inputs/src/Node.cpp:76]: The line is not created because we check that outside the builder

File: sources/Inputs/src/Node.cpp (hunk 6 above)
Comment (lecourtoisflo) [reply sources/Inputs/src/Node.cpp:74]: OK

File: sources/Outputs/src/Dyd.cpp (hunk 7 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Dyd.cpp:164]: Formatting error which added a comment, this braket does not correspond to a namespace

File: tests/inputs/TestNode.cpp (hunk 8 above)
Comment (lecourtoisflo) [reply tests/inputs/TestNode.cpp:34]: OK

Comment (lecourtoisflo) [reply tests/inputs/TestNode.cpp:34]: These are covered even without new tests

File: sources/Inputs/src/Node.cpp (hunk 5 above)
Comment (rosiereflo) [reply sources/Inputs/src/Node.cpp:76]: could you please add an assert ?

File: tests/inputs/TestNode.cpp (hunk 8 above)
Comment (rosiereflo) [reply tests/inputs/TestNode.cpp:34]: What is the latest code coverage?

Comment (lecourtoisflo) [reply tests/inputs/TestNode.cpp:34]: 88% in total, almost 100% for inputs

File: sources/Inputs/src/Node.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Inputs/src/Node.cpp:76]: OK

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**